target-platforms/packages/portal-base/package.json

```diff
@@ -0,0 +1,23 @@
+{
+	"dependencies": {
+		"@babel/core": "^7.14.3",
+		"@babel/preset-env": "^7.14.2",
+		"@babel/preset-react": "^7.13.13",
+		"babel-preset-liferay-standard": "^2.26.0",
+		"liferay-npm-bundler-loader-copy-loader": "^2.26.0",
+		"liferay-npm-bundler-loader-babel-loader": "^2.26.0",
+		"liferay-npm-bundler-plugin-exclude-imports": "^2.26.0",
+		"liferay-npm-bundler-plugin-inject-imports-dependencies": "^2.26.0",
+		"liferay-npm-bundler-plugin-inject-peer-dependencies": "^2.26.0",
+		"liferay-npm-bundler-plugin-namespace-packages": "^2.26.0",
+		"liferay-npm-bundler-plugin-replace-browser-modules": "^2.26.0",
+		"liferay-npm-bundler-plugin-resolve-linked-dependencies": "^2.26.0"
```

Comment:
What if we change a version of these dependencies only for a single target platform? Wouldn't we want target platforms to not share?

It seems like portal-base would be tied to each individual version of DXP but I may be misunderstanding the purpose of this package.

Reply:
I think I am just slightly confused on the need/purpose of portal-base. It seems like it might be more of a hassle to publish for each new target platform and then making sure old target platforms don't rely on a newer portal-base

Reply:
> What if we change a version of these dependencies only for a single target platform? Wouldn't we want target platforms to not share?

If that case happens we will need another refactor to split up again. The truth is that we will need to evolve the architecture of target platforms as time passes by. But since they are like a black box for the user and and their only point of contact is the `liferay` npm binary we can change them as much as we want/need.

> It seems like portal-base would be tied to each individual version of DXP but I may be misunderstanding the purpose of this package.

It depends. As with `npm-scripts` we can have one pseudo/real branch per DXP version or -better- try to rule them all with one single package (as we wanted to do in the future with `npm-scripts`). I would say we will be able to use the same package for every DXP version if we account for that possibility. In the end, if we upgrade the build to do more things or use newer versions of devtools, that doesn't mean we cannot target older DXP versions. See the bundler: people have been always able to use the latest version for any DXP version.

> I think I am just slightly confused on the need/purpose of portal-base. It seems like it might be more of a hassle to publish for each new target platform and then making sure old target platforms don't rely on a newer portal-base

In fact, I started target platforms with the opposite model in mind (each platform has its own code), but since Javi said Brian would need to publish platforms (not us) we cannot rely on that model because we have split responsibilities:

1) Brian will define when the platform is released and will run a script we provide to release them
2) We need to have a way to fix problems in released platforms without having to wait days for someone to release something

Per 1, we want to provide a `create-platform` script that is fast, simple and fail-safe. Something we are far from with the first `create-platform` script I created.

Per 2, we need to be able to publish new versions of portal-base, then tell the customers to upgrade their `yarn.lock` files to update their platform (without any need to release a new platform).

To sum up:

- The target platform package only has the configuration to define how a project for a given DXP version is built and what the imports are (immutable, at least the imports part)
- The portal-base package implements all the logic and can be upgraded without touching target platforms. We "simply" need to take care of making it backwards and forward compatible (something that I think won't be so difficult).

Reply:
In any case, we are making a lot of assumptions here, upfront, and some may prove wrong in the future. This is an invariant in all tools development :roll_eyes:, but the good news are that, as long as you define a coherent model and control the surface of the tools (not exposing too much API) to be able to change underlying things it usually works and is maintainable. At least that has been my experience with the bundler and the AMD loader...

Reply:
Ah I see, thanks for the explanation, that helps clarify and gives good context.

> Brian would need to publish platforms (not us) we cannot rely on that model because we have split responsibilities

I hadn't totally thought about this side of things, so you are right in that it helps give us a bit more freedom to publish fixes to portal-base without having to rely bchan publishing.



> In any case, we are making a lot of assumptions here, upfront, and some may prove wrong in the future. This is an invariant in all tools development 🙄, but the good news are that, as long as you define a coherent model and control the surface of the tools (not exposing too much API), to be able to change underlying things, it usually works and is maintainable.

Yep, totally makes sense!